retry/src/main/java/com/iluwatar/retry/App.java

```diff
@@ -0,0 +1,121 @@
+ * operations that our application performs involving remote systems. The calling code should remain
+ * decoupled from implementations.
+ *
+ * <p>{@link FindCustomer} is a business operation that looks up a customer's record and returns
+ * its ID. Imagine its job is performed by looking up the customer in our local database and
+ * returning its ID. We can pass {@link CustomerNotFoundException} as one of its {@link
+ * FindCustomer#FindCustomer(java.lang.String, com.iluwatar.retry.BusinessException...) constructor
+ * parameters} in order to simulate not finding the customer.
+ *
+ * <p>Imagine that, lately, this operation has experienced intermittent failures due to some weird
+ * corruption and/or locking in the data. After retrying a few times the customer is found. The
+ * database is still, however, expected to always be available.  While a definitive solution is
+ * found to the problem, our engineers advise us to retry the operation a set number of times with a
+ * set delay between retries, although not too many retries otherwise the end user will be left
+ * waiting for a long time, while delays that are too short will not allow the database to recover
+ * from the load.
+ *
+ * <p>To keep the calling code as decoupled as possible from this workaround, we have implemented
+ * the retry mechanism as a {@link BusinessOperation} named {@link Retry}.
+ *
+ * @author George Aristy (george.aristy@gmail.com)
+ * @see <a href="https://docs.microsoft.com/en-us/azure/architecture/patterns/retry">Retry pattern
+ *     (Microsoft Azure Docs)</a>
+ */
+public final class App {
+  private static final Logger LOG = LoggerFactory.getLogger(App.class);
+  private static BusinessOperation<String> op;
+
+  /**
+   * Entry point.
+   *
+   * @param args not used
+   * @throws Exception not expected
+   */
+  public static void main(String[] args) throws Exception {
+    noErrors();
+    errorNoRetry();
+    errorWithRetry();
+    errorWithRetryExponentialBackoff();
+  }
+
+  private static void noErrors() throws Exception {
+    op = new FindCustomer("123");
+    op.perform();
+    LOG.info("Sometimes the operation executes with no errors.");
+  }
+
+  private static void errorNoRetry() throws Exception {
+    op = new FindCustomer("123", new CustomerNotFoundException("not found"));
+    try {
+      op.perform();
+    } catch (CustomerNotFoundException e) {
+      LOG.info("Yet the operation will throw an error every once in a while.");
+    }
+  }
+
+  private static void errorWithRetry() throws Exception {
+    final Retry<String> retry = new Retry<>(
+        new FindCustomer("123", new CustomerNotFoundException("not found")),
+        3,  //3 attempts
+        100, //100 ms delay between attempts
+        e -> CustomerNotFoundException.class.isAssignableFrom(e.getClass())
+    );
+    op = retry;
+    final String customerId = op.perform();
+    LOG.info(String.format(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

This code uses '%s' to format an integer: 'attempts' expression. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 101. Consider formatting this integer with '%d' instead.